### PR TITLE
Add geolocation on tax button #64 #51

### PR DESCRIPTION
**Pull request by miisaka**
**User Story #51**
**Tasks:** #64 #62

### WHAT kind of change does this PR introduce?
Actual geolocation for the tax button that will find the current location of the user. Location gets updated every 30000ms or 1000m and when the user presses on the "$" tab, a prompt will request for the user to grant access to the location service of the device.  

When the location service is enabled, pressing on the tax button will get the user's geolocation and calculate the tax depending on the province within Canada the user is located in currently. If the location service is disabled, then a modal with all of Canada's province will spawn and the user can select the province they want to calculate their taxes with.

device-2018-03-08-022049

...

### HOW is this accomplished?
Using the location service of the device that gets the location via GPS. Implemented using Android's LocationManager library.
...

### Checklist
- [x] Docs have been added / updated to reflect the changes
- [x] I have reviewed and tested the changes :white_check_mark:
- [x] I have squashed my commits to have a reasonable amount of commits
- [x] All of the code I have written adhere to the coding conventions outlined in the coding conventions section of the wiki

### CODE REVIEW

**Comment by miisaka** [comment]
Some major refactoring will need to be done and unit tests will need to be added for this feature

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 1]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
         btn_tip.setOnClickListener { true } // TODO : Implement feature and connect
         result.setOnLongClickListener { copyToClipboard(result.value); true }
+        btn_taxes.setOnClickListener({ view -> taxLocationStrat() })
 
         AutofitHelper.create(result)
 
     }
 
-     override fun spawnTaxModal() {
+    override fun onLocationChanged(location: Location) {
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Is this method every called? Also consider making this method private, we shouldn't be exposing this functionality.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 2]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+            gps_enabled = locationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)
+        } catch(e: Exception) {}
+
+        //if location service is enabled then retrieve the latest location and performs tax rate with that location
+        if(gps_enabled) {
+            //Checks if location service has been granted again
+            if (ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                ActivityCompat.requestPermissions(this, arrayOf(android.Manifest.permission.ACCESS_FINE_LOCATION), GPS_REQUEST_CODE);
+            }
+            var lastLocation = locationManager!!.getLastKnownLocation(LocationManager.GPS_PROVIDER)
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Do we need lastLocation? Can't we just store the last known location in currentLongitude and currentLatitude and update them when we the geoloc is refreshed?

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 3]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if(requestCode == GPS_REQUEST_CODE) {
+            if(grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                Log.e("PermissionTag","Permission Granted (if 0 then is granted): " + grantResults[0])
+            }
+        }
+    }
+
+    fun taxLocationStrat() {
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: This function can be private since it's helper method internal to this activity.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 4]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+        var gps_enabled : Boolean = false;
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: **Style catch:** Variables in Kotlin should use camelCase. The semicolon at the end of this line can be removed

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 5]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+            provinceAbbreviationMap.put("QC", "Quebec")
+            provinceAbbreviationMap.put("NB", "New Brunswick")
+            provinceAbbreviationMap.put("NS", "Nova Scotia")
+            provinceAbbreviationMap.put("PE", "Prince Edward Island")
+            provinceAbbreviationMap.put("NL", "Newfoundland and Labrador")
+            provinceAbbreviationMap.put("NT", "Northwest Territories")
+            provinceAbbreviationMap.put("NU", "Nunavut")
+            provinceAbbreviationMap.put("YT", "Yukon")
+
+            calc.performTaxing(provinceAbbreviationMap.get(province).toString())
```
Comment (izconcept) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Shouldn't this part(lines 160-176) be performed within the try block on line 137?

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 6]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+    override fun onLocationChanged(location: Location) {
+        currentLatitude = location.latitude
+        currentLongitude = location.longitude
+        val message = String.format("New Location \n Longitude: %1\$s \n Latitude: %2\$s",
+                location.longitude, location.latitude
+        )
+        Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
+    }
+
+    override fun onStatusChanged(s:String, i:Int, b:Bundle) {}
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Why is this overriden if it isnt being used ?

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 3 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: I agreed

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 4 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Good styling catch

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 7]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+            try {
+                 province = addresses.get(0).adminArea
+            } catch (e: Exception) {
+                e.printStackTrace()
+                Toast.makeText(applicationContext, "Unable to recognize your location, please choose one of the following provinces", Toast.LENGTH_SHORT).show()
+                spawnTaxModal()
+                return
+            }
+
+//            //TODO: refactor this part to make it cleaner
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Good call on that

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 5 above)
Comment (almiche) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: I dont see why

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt [hunk 8]
```diff
@@ -86,14 +86,10 @@ class MoneyCalculatorImpl(calculator: Calculator,taxCalculator: TaxCalculator, v
...
-        if(false){
-
-        }else{
-            tCallback!!.spawnTaxModal()
-        }
-    }
+    //TODO: Refactor this
+//    fun calculateTax(){
+//        tCallback!!.spawnTaxModal()
+//    }
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: Uncomment this line

File: app/src/main/kotlin/com/simplemobiletools/calculator/operation/TaxOperation.kt [hunk 9]
```diff
@@ -15,6 +15,9 @@ class TaxOperation(var value: Double, var location: String)  {
         taxesMap.put("Nova Scotia", 0.15);
         taxesMap.put("Prince Edward Island", 0.15);
         taxesMap.put("Newfoundland and Labrador", 0.15);
+        taxesMap.put("Northwest Territories", 0.05);
+        taxesMap.put("Nunavut", 0.05);
+        taxesMap.put("Yukon", 0.05);
```
Comment (almiche) [inline app/src/main/kotlin/com/simplemobiletools/calculator/operation/TaxOperation.kt]: Nice catch we're all about being inclusive

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 10]
```diff
@@ -3,33 +3,46 @@ package com.simplemobiletools.calculator.activities
...
+class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator, LocationListener {
 
     private var storedTextColor = 0
     private var vibrateOnButtonPress = true
     private var storedUseEnglish = false
     private var taxDialog:AlertDialog.Builder? = null
+    private var locationManager: LocationManager? = null
+    private var GPS_REQUEST_CODE = 101
+    private var MINIMUM_TIME_BETWEEN_UPDATES : Long = 300000L //in ms
+    private var MINIMUM_DISTANCE_CHANGE_FOR_UPDATES : Float = 100f //in meters
```
[the remark this answers is not in the document]
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Makes sense, will change it to 1000m.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 1 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: This method is called whenever a new location is being updated by line 66. This method is also being implemented from the interface LocationListener therefore can not be private

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 6 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Since this class implements the LocationManager interface therefore i have to implement all the interface's method

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 3 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: I agree. Will change it to private

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 4 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Great catch! This totally slipped me completely! What would I do without you guys!!!!! You are all the best!!!1!!

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 2 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: We could but since this is methodology of getting from the last known location is more foolproof as it does not rely on time nor on the user changing location before both currentLongitude and currentLatitude to be updated. Furthermore this is not adding any further dependencies as it is just calling the getLastKnownLocation from the locationManager so I don't think this needs to be changed

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt [hunk 11]
```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
...
+
+            if (lastLocation != null) {
+                currentLongitude = lastLocation.longitude
+                currentLatitude = lastLocation.latitude
+                var message: String = String.format("Current Location \n Longitude: " + lastLocation.longitude + " \n Latitude: " + lastLocation.latitude)
+                Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
+            }
+
+            val geocoder = Geocoder(applicationContext, Locale.getDefault())
+            val addresses: List<Address> = geocoder.getFromLocation(currentLatitude, currentLatitude, 1)
```
[the remark this answers is not in the document]
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: nice catch, my silly mistake. That may increase the performance but I don't it is worth the effort as the address List collected here is only of a size of 1 therefore it will be fetched very quickly enough and multithreading this part will result in marginal gain of time

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 5 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: why should this part be part of a try-catch block? There won't be any errors thrown here but it is true that this part l.106-174 should probably be put in a separate file from this MoneyActivity class

File: app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt (hunk 8 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/helpers/MoneyCalculatorImpl.kt]: This method as well as the TaxCalculator should be removed completely. I will refactor this in my next commit.

File: app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt (hunk 7 above)
Comment (miisaka) [reply app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt]: Refactoring will be done in my next commit.

**Comment by jusleg** [comment]
make sure you commit message follow the proper formatting

**Comment by miisaka** [comment]
@jusleg forgot to include issue # on the comment message whuuuuups. Added comments to reference the issues on the commits

File: app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt [hunk 12]
```diff
@@ -90,7 +112,68 @@ class MoneyCalculatorImplTest {
     }
 
     @Test
-    fun calculateTax(){
+    fun calculateTaxBritishColumbia(){
+        calc.addDigit(2)
```
Comment (almiche) [inline app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt]: This should be refactored into a method since it's being done 10 times

Comment (miisaka) [reply app/src/test/java/com/simplemobiletools/calculator/MoneyCalculatorImplTest.kt]: What should be refactored? Each test is testing for a different province tax rate

**Comment by almiche** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

✅ Branch has an acceptable amount of commits with descriptive names

✅PR description matches the code outlined a has mentions to issues

✅This PR attempts to implement a specific feature 

Status: **Approved**
Comments: All looks well fixes were made

**Comment by 0matleb2** [comment]
✅ Styling is adequate and follows coding conventions outlined by the [Kotlin Foundation](https://kotlinlang.org/docs/reference/coding-conventions.html)

✅ New methods and functionalities have appropriate tests (testing edge cases + good coverage)

✅ Branch has an acceptable amount of commits with descriptive names

✅PR description matches the code outlined a has mentions to issues

✅This PR attempts to implement a specific feature 

Status: **Approved**
Comments: Look great. Good job on this PR.
